flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/queue/StreamElementQueue.java

```diff
@@ -19,68 +19,56 @@
+import org.apache.flink.streaming.runtime.streamrecord.StreamElement;
 
-import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
 
 /**
- * Interface for blocking stream element queues for the {@link AsyncWaitOperator}.
+ * Interface for stream element queues for the {@link AsyncWaitOperator}.
  */
 @Internal
-public interface StreamElementQueue {
+public interface StreamElementQueue<OUT> {
 
 	/**
-	 * Put the given element in the queue if capacity is left. If not, then block until this is
-	 * the case.
+	 * Tries to put the given element in the queue. This operation succeeds if the queue has capacity left and fails if
+	 * the queue is full.
 	 *
-	 * @param streamElementQueueEntry to be put into the queue
-	 * @param <T> Type of the entries future value
-	 * @throws InterruptedException if the calling thread has been interrupted while waiting to
-	 * 	insert the given element
-	 */
-	<T> void put(StreamElementQueueEntry<T> streamElementQueueEntry) throws InterruptedException;
-
-	/**
-	 * Try to put the given element in the queue. This operation succeeds if the queue has capacity
-	 * left and fails if the queue is full.
+	 * <p>This method returns a handle to the inserted element that allows to set the result of the computation.
 	 *
-	 * @param streamElementQueueEntry to be inserted
-	 * @param <T> Type of the entries future value
-	 * @return True if the entry could be inserted; otherwise false
-	 * @throws InterruptedException if the calling thread has been interrupted while waiting to
-	 * 	insert the given element
+	 * @param streamElement the element to be inserted.
+	 * @return A handle to the element if successful or {@link Optional#empty()} otherwise.
 	 */
-	<T> boolean tryPut(StreamElementQueueEntry<T> streamElementQueueEntry) throws InterruptedException;
+	Optional<ResultFuture<OUT>> tryPut(StreamElement streamElement);
 
 	/**
-	 * Peek at the head of the queue and return the first completed {@link AsyncResult}. This
-	 * operation is a blocking operation and only returns once a completed async result has been
-	 * found.
+	 * Emits one completed element from the head of this queue into the given output.
 	 *
-	 * @return Completed {@link AsyncResult}
-	 * @throws InterruptedException if the current thread has been interrupted while waiting for a
-	 * 	completed async result.
+	 * <p>Will not emit any element if no element has been completed (check {@link #hasCompletedElements()} before entering
+	 * any critical section).
+	 *
+	 * @param output the output into which to emit
 	 */
-	AsyncResult peekBlockingly() throws InterruptedException;
+	void emitCompletedElement(TimestampedCollector<OUT> output);
 
 	/**
-	 * Poll the first completed {@link AsyncResult} from the head of this queue. This operation is
-	 * blocking and only returns once a completed async result has been found.
+	 * Checks if there is at least one completed element.
```

Comment:
This description is not very accurate for the ordered case. For order case, it must be the head element completed in queue.